dynamic_scenario_builder.go

```diff
@@ -0,0 +1,61 @@
+package yoti
+
+import (
+	"encoding/json"
+)
+
+// DynamicScenarioBuilder builds a Dynamic Scenario
+type DynamicScenarioBuilder struct {
+	scenario DynamicScenario
+}
+
+// DynamicScenario represents a dynamic scenario
+type DynamicScenario struct {
+	policy           DynamicPolicy
+	extensions       []interface{}
+	callbackEndpoint string
+}
+
+// New initializes the state of a DynamicScenarioBuilder before its use
+func (builder *DynamicScenarioBuilder) New() *DynamicScenarioBuilder {
+	builder.scenario.policy = (&DynamicPolicyBuilder{}).New().Build()
+	builder.scenario.extensions = make([]interface{}, 0)
+	builder.scenario.callbackEndpoint = ""
+	return builder
+}
+
+// WithPolicy attaches a DynamicPolicy to the DynamicScenario
+func (builder *DynamicScenarioBuilder) WithPolicy(policy DynamicPolicy) *DynamicScenarioBuilder {
+	builder.scenario.policy = policy
+	return builder
+}
+
+// WithExtension adds an extension to the scenario
+func (builder *DynamicScenarioBuilder) WithExtension(extension interface{}) *DynamicScenarioBuilder {
+	builder.scenario.extensions = append(builder.scenario.extensions, extension)
+	return builder
+}
+
+// WithCallbackEndpoint sets the callback URL
+func (builder *DynamicScenarioBuilder) WithCallbackEndpoint(endpoint string) *DynamicScenarioBuilder {
+	builder.scenario.callbackEndpoint = endpoint
+	return builder
+}
+
+// Build constructs the DynamicScenario
+func (builder *DynamicScenarioBuilder) Build() DynamicScenario {
+	return builder.scenario
+}
+
+// MarshalJSON ...
```

Comment:
Some linters (specifically the one build into vim-go) complains if you have an exported function without a doc comment, but the MarshalJSON function is built into how encoding/json json.Marshal works. I couldn't think of a good boilerplate comment to put here.

Reply:
Yes we should always have comments on exported functions. 

I think we should have always have some text, even if it can seem redundant. We could use something like 
> // MarshalJSON returns m as the JSON encoding of m.

(From https://golang.google.cn/src/encoding/json/stream.go - L256)